File: go/LeetCode/findTargetSumWays.go
package main

import "fmt"

func main() {
	nums := []int{1,1,1,1,1}
	//nums := []int{1}
	fmt.Println(findTargetSumWays2(nums, 3))
}


func findTargetSumWays(nums []int, target int) (count int) {
	var dfs func(int,int)
	dfs = func(index int, sum int) {
		if index == len(nums) {
			if sum == target {
				count++
			}
		}else {
			dfs(index+1, sum+nums[index])
			dfs(index+1, sum-nums[index])
		}
	}
	dfs(0,0)
	return count
}



func findTargetSumWays1(nums []int, target int) int {
	sum:=0
	for i:=0;i<len(nums);i++ {
		sum+=nums[i]
	}
	diff := sum-target
	if diff<0 || diff%2!=0 {
		return 0
	}
	mins := diff/2
	dp:=make([][]int, len(nums)+1)
	for i:= range dp{
		dp[i] = make([]int, mins+1)
	}
	dp[0][0]=1

	for i:=1;i<=len(nums);i++ {
		num := nums[i-1]
		for j:=0;j<=mins;j++ {
			dp[i][j] = dp[i-1][j]
			if j>=num {
				dp[i][j]+=dp[i-1][j-num]
			}
		}
	}
	return dp[len(nums)][mins]

}

func findTargetSumWays2(nums []int, target int) int {
	sum:=0
	for _,v := range nums{
		sum+=v
	}
	diff := sum-target
	if diff<0 || diff%2!=0 {
		return 0
	}
	mins := diff/2
	dp := make([]int, mins+1)
	dp[0]=1
	for _,v := range nums{
		for j:=mins;j>=v;j-- {
			dp[j] += dp[j-v]
		}
	}
	return dp[mins]
}

File: python3/LeetCode/findTargetSumWays.py
from typing import List


count=0
class Solution:
    def findTargetSumWays(self, nums:List[int], target:int) ->int:
        global count
        self.dfs(nums, target, 0,0)
        return count


    def dfs(self, nums:List[int], target:int, index:int, sum:int):
        if index == len(nums):
            if sum == target:
                global count
                count+=1
        else:
            self.dfs(nums, target, index+1,sum+nums[index])
            self.dfs(nums, target, index+1,sum-nums[index])


    def findTargetSumWays1(self, nums:List[int], target:int) ->int:
        sum = 0
        for v in nums:
            sum+=v
        diff = sum - target
        if diff<0 or diff%2!=0:
            return 0
        mins = diff//2
        dp = [[0]*(mins+1) for i in range(len(nums)+1)]
        dp[0][0] = 1
        for i in range(1, len(nums)+1):
            num = nums[i-1]
            for j in range(mins+1):
                dp[i][j] = dp[i-1][j]
                if j>= num:
                    dp[i][j] += dp[i-1][j-num]

        return dp[len(nums)][mins]

    def findTargetSumWays2(self, nums:List[int], target:int) ->int:
        sum =0
        for v in nums:
            sum+=v
        diff = sum -target
        if diff<0 or diff%2!=0:
            return 0
        mins = diff//2
        dp = [0]*(mins+1)
        dp[0]=1
        for v in nums:
            for j in range(mins,v-1,-1):
                dp[j] += dp[j-v]
        return dp[mins]

if __name__ == '__main__':
    # nums=[7,9,3,8,0,2,4,8,3,9]
    nums=[1,1,1,1,1]
    print(Solution().findTargetSumWays2(nums, 3))